tests/load/transformations/test_incremental_transforms.py

```diff
@@ -0,0 +1,237 @@
+        destination="duckdb",
+        dataset_name="incremental_dataset",
+        dev_mode=True,
+    )
+    return pipeline
+
+
+@dlt.resource(table_name="items")
+def first_load() -> Iterator[TDataItem]:
+    yield from [
+        {"id": 1},
+        {"id": 2},
+        {"id": 3},
+    ]
+
+
+@dlt.resource(table_name="items")
+def inc_load() -> Iterator[TDataItem]:
+    yield from [
+        {"id": 4},
+        {"id": 5},
+        {"id": 6},
+    ]
+
+
+EXPECTED_TRANSFORMED_DATA_FIRST_LOAD = [
+    {"id": 1, "double_items": 2},
+    {"id": 2, "double_items": 4},
+    {"id": 3, "double_items": 6},
+]
+
+EXPECTED_TRANSFORMED_DATA_SECOND_LOAD = [
+    {"id": 1, "double_items": 2},
+    {"id": 2, "double_items": 4},
+    {"id": 3, "double_items": 6},
+    {"id": 4, "double_items": 8},
+    {"id": 5, "double_items": 10},
+    {"id": 6, "double_items": 12},
+]
+
+
+def _assert_transformed_data(inc_p: dlt.Pipeline, expected_data: list[Any]) -> None:
+    found_data = (
+        inc_p.dataset()
+        .transformed_items.select("id", "double_items")
+        .df()
+        .to_dict(orient="records")
+    )
+
+    # make all dict keys lowercase for our favorite snowflake
+    found_data = [{k.lower(): v for k, v in d.items()} for d in found_data]
+    sorted_found_data = sorted(found_data, key=lambda x: x["id"])
+
+    assert sorted_found_data == expected_data
+
+
+@pytest.mark.essential
+@pytest.mark.parametrize(
+    "destination_config",
+    transformation_configs(),
+    ids=lambda x: x.name,
+)
+@pytest.mark.parametrize("transformation_type", ["python", "sql"])
+def test_state_based_incremental_transform(
+    destination_config: DestinationTestConfiguration, transformation_type: Any
+) -> None:
+    """
+    Here we demonstrate how to use the resource state for transform incrementals
+    """
+    # get pipelines and populate fruit pipeline
+    inc_p, dest_p = setup_transformation_pipelines(destination_config, transformation_type)
+
+    @dlt.transformation(
+        transformation_type=transformation_type,
+        write_disposition="append",
+    )
+    def transformed_items(dataset: SupportsReadableDataset[Any], last_loaded_load_id: str) -> Any:
+        # get last stored processed load id
+        LAST_PROCESSED_LOAD_ID = "last_processed_load_id"
+        last_processed_load_id = dlt.current.resource_state().get(LAST_PROCESSED_LOAD_ID, "0")
+        items_table = dataset.items
+
+        max_load_id = list(
+            items_table._dlt_load_id.max().df().to_dict(orient="records")[0].values()
+        )[0]
+        dlt.current.resource_state()[LAST_PROCESSED_LOAD_ID] = max_load_id
+
+        # return filtered transformation
+        return items_table.filter(
+            items_table._dlt_load_id > last_processed_load_id,
+            items_table._dlt_load_id <= last_loaded_load_id,
+        ).mutate(double_items=items_table.id * 2)
+
+    # first round
+    inc_p.run(first_load())
+    last_loaded_load_id = list(
```

Comment:
why you do it here? that should be in the resource using `dlt.current.dataset()`
maybe it would be good to have `scalar()` after all. how we get value is ugly
also maybe our dataset should have an info if it s new or not (new == does not exist on destination)

Reply:
I don't think I agree. We need to determine the last fully completed load of the origin dataset before the pipeline is started (or alternatively at the beginning of a source where it is then forwarded into the resources), because if we have multiple transformation resources and they all do this query independently, there may be the case that there is another completed load that finishes just during the extraction phase, and the transformations of this load will work with different load_ids as the end of the range.

Reply:
That's why I liked your idea from the other PR where you suggested instantiating a dataset that knows from which and until which other load_id the records are "new". Then all transformations will run with the same range.

Reply:
PS: I added the scalar